### PR TITLE
Add unlink reasons to unlink function

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
include unlink reasons when posting an unlink to the adaptor

#### Ticket

[CBO-1270](https://dsdmoj.atlassian.net/browse/CBO-1270)

#### Why
MLRA/MAATDB audits reasons for unlinking a defendant
from a given MAAT reference. We therefore need to
supply this.

It is intended to eventually use the adpator to lookup
unlinking reasons from MLRA via MAAT API and the adaptor
but, for now, we are storing the 7 reasons in VCD.

--------

#### TODO (wip)

 - [x] add table, model
 - [x] add form for unink reasons
 - [x] handle other reasons text submission
 - [x] submit unlink reason with unlink request
 - [x] meet designs
 - [x] fully spec and test
 - [x] amend adaptor to add unlink_other_reason_text - in review
 - [x] resolve integer string issue for code sending to adaptor (request/feature test differences with params)
 - [x] refactor defendant controller
 - [x] show hide other reason box
 - [x] display errors and links as if its a model

### CODE REVIEW

File: app/webpack/stylesheets/components/_panels.scss [hunk 1]
```diff
@@ -0,0 +1,30 @@
+.supplier-numbers-wrapper .panel .form-group,
```
Comment (naseberry) [inline app/webpack/stylesheets/components/_panels.scss:1]: @jsugarman you can remove this altogether and use the Details component instead: 
https://design-system.service.gov.uk/components/details/

File: app/views/laa_references/_unlink.html.haml [hunk 2]
```diff
@@ -1,4 +1,20 @@
-= link_to t('laa_reference.unlink.label'), remove_link_defendant_path(@defendant.arrest_summons_number), method: :delete, class: 'govuk-link'
+%details
+  %summary
+    %span.summary
+      = t('laa_reference.unlink.detail_link')
+  / .govuk-inset-text
+  .panel.panel-border-narrow
```
Comment (naseberry) [inline app/views/laa_references/_unlink.html.haml:6]: Replace with the Details component.

```
%details.govuk-details{ 'data-module': 'govuk-details' }
  %summary.govuk-details__summary
    %span.govuk-details__summary-text
      = t('laa_reference.unlink.detail_link')
  .govuk-details__text
```

Comment (jsugarman) [reply app/views/laa_references/_unlink.html.haml:6]: brill, thanks

File: app/webpack/stylesheets/components/_panels.scss (hunk 1 above)
Comment (jsugarman) [reply app/webpack/stylesheets/components/_panels.scss]: thats done now

File: app/views/laa_references/_unlink.html.haml (hunk 2 above)
Comment (jsugarman) [reply app/views/laa_references/_unlink.html.haml]: done

File: db/seeds.rb [hunk 3]
```diff
@@ -1,41 +1,15 @@
 # frozen_string_literal: true
 
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
+# Seed all lookup data
```
Comment (jsugarman) [inline db/seeds.rb:3]: seed all data (not just lookup data)

File: app/webpack/stylesheets/components/_panels.scss (hunk 1 above)
Comment (jsugarman) [reply app/webpack/stylesheets/components/_panels.scss]: removed the commit with this css and manifest change

File: app/helpers/govuk_design_system_helper.rb [hunk 4]
```diff
@@ -21,8 +21,28 @@ def govuk_breadcrumb_builder
     GdsDesignSystemBreadcrumbBuilder
   end
 
+  def govuk_detail(summary_text = nil, open = false, tag_options = {}, &block)
```
[the remark this answers is not in the document]
Comment (jsugarman) [reply app/helpers/govuk_design_system_helper.rb:24]: cool, feel free to to open a quick PR extending that once this is in
i spent a fair bit of time trying to test the open/closed functionality in feature test but hit so many problems with capybara i quit. Worth looking into thought - to extend the existing custom rspec matcher to handle test opened and closed 

this is where i got to
```
#spec/support/capybara_extensions.rb
    def has_govuk_detail_summary?(text, options = {})
      check_open = options.key?(:open)
      open_type = options.delete(:open)

      detail = find(detail_summary_selector, options.merge(text: text))

      return detail.present? unless check_open

      case open_type
      in false | :hidden | :closed
        ap "checking closed"
        has_no_selector?("detail.govuk-details[open]")
      in true | :open | :expanded
        ap "checking open"
        has_selector?("detail.govuk-details[open]")
      else
        raise ArgumentError, "invalid open option #{open_type || 'nil'}. value should be one of (true || :open || :expanded) or (false || :hidden)"
      end
    end
```

**Comment by naseberry** [comment]
@jsugarman could you make the history cleaner by squashing some of these commits?

File: app/helpers/govuk_design_system_helper.rb (hunk 4 above)
Comment (kmahern) [inline app/helpers/govuk_design_system_helper.rb:24]: This is great. In order to be consistent we should use this for the govuk_details I added as part of unlinking without a maat id (although not as part of this pr).

File: app/views/defendants/_defendant.html.haml [hunk 5]
```diff
@@ -25,13 +23,4 @@
...
-
-= render('representation_orders', locals: @defendant) if @defendant.linked?
-
-= render 'offences', locals: @defendant
-
-= render 'laa_references/unlink', locals: @defendant if @defendant.linked?
-
-= render('laa_references/form', locals: @defendant) unless @defendant.linked?
-
+= link_to 'View case', prosecution_case_path(@defendant.prosecution_case_reference), class: 'govuk-button govuk-button--secondary'
```
Comment (kmahern) [inline app/views/defendants/_defendant.html.haml:26]: Missing newline at end of file?

File: app/webpack/packs/unlink.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+function Unlinking () {
+  // rely on `name`` because id changes in error state
+  //
+  const unlinkReasonCode = document.getElementsByName('unlink_attempt[reason_code]')[0]
+  const unlinkOtherReasonText = document.getElementsByName('unlink_attempt[other_reason_text]')[0]
+  const otherReasonCode = '7'
+
```
Comment (naseberry) [inline app/webpack/packs/unlink.js:7]: @jsugarman you may wish to add `if (!unlinkReasonCode) { return false }` before line 8.

The return statement executes if the element is undefined, it therefore stops the rest of the code from running, this helps performance and reduces errors

**Comment by jsugarman** [comment]
> @jsugarman could you make the history cleaner by squashing some of these commits?

yes, i do that after a review usually in  case i need to add stuff

**Pull request merged**